### ISSUE TITLE
Error writing to FileGDB "The spatial index grid size is invalid"

### ISSUE THREAD

**Issue by keithamoss** [Toblerity/Fiona#388]
I've hit a bit of an interesting edge case with creating FileGDBs using Fiona.

**tl;dr** I have a dataset of MultiPolygons of widely ranging sizes that I need to write to a FileGDB with a single layer. When the first feature in the dataset is several orders of magnitude smaller than the largest feature Fiona encounters an `ERROR:Fiona:CPLE_AppDefined in Failed at writing Row to Table in CreateFeature. (The spatial index grid size is invalid.)` error.

It appears that the first feature written is used to setup the internal spatial domain of a FileGDB[1], though I've not been able to find any documentation from ESRI, the OpenFileGDB reverse engineering, or through general research that confirms this. It is documented[2] that exceeding this domain causes issues in spatial indexing though.

In practice however, we see that the first feature does appear to dictate the bounds of the internal spatial domain.

See https://gist.github.com/keithamoss/e4c15cb8ecef143901b6131baad91bbd for the three test datasets.

**NB: I've used `fio` to debug the issue, but in practice this is part of a larger Flask+Fiona+GeoPandas app.**
### **small-geom-first.json: A GeoJSON dataset with two polygons (small polygon first, large polygon second)**

`fio dump small-geom-first.json | fio load test.gdb --driver FileGDB --src-crs EPSG:4326`

**Creates FileGDB containing the small polygon, then aborts.**

```
WARNING:Fiona:CPLE_NotSupported in dataset /app/tmp/filegdb-bug/test.gdb does not support layer creation option ENCODING
ERROR:Fiona:CPLE_AppDefined in Failed at writing Row to Table in CreateFeature. (The spatial index grid size is invalid.)
ERROR:fio:Exception caught during processing
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/fiona/fio/load.py", line 103, in load
    dst.writerecords(source)
  File "/usr/local/lib/python2.7/site-packages/fiona/collection.py", line 327, in writerecords
    self.session.writerecs(records, self)
  File "fiona/ogrext.pyx", line 1038, in fiona.ogrext.WritingSession.writerecs (fiona/ogrext2.c:17532)
RuntimeError: Failed to write record: {u'geometry': {u'type': u'Polygon', u'coordinates': [[[127.29999999999998, -29.01], [119.49, -32.28], [118.43, -27.1], [125.9, -21.44], [127.29999999999998, -29.01]]]}, u'type': u'Feature', u'id': u'1', u'properties': {}}
Aborted!
```
### **first-geom-wa-extent.json: A GeoJSON dataset with three polygons (polygon with Western Australia's extent first, small polygon second, large polygon third)**

`fio dump /app/tmp/filegdb-bug/first-geom-wa-extent.json | fio load /app/tmp/filegdb-bug/test.gdb --driver FileGDB --src-crs EPSG:4326`

**Success! Creates FileGDB containing both records.**
### **first-geom-empty.json: A GeoJSON dataset with three polygons (polygon with empty geometry first, small polygon second, large polygon third)**

This fails via `fio dump | fio load` with an `ValueError: Record's geometry type does not match collection schema's geometry type: u'Polygon' != u'MultiPolygon'` error, but does successfully create a FileGDB if run through Fiona in Python (see code below).
## ogr2ogr

Running all of these GeoJSON files through `ogr2ogr` will successfully create a FileGDB.

I'm unsure (hence filing this bug) if the issue here lays with Fiona's interfacing with GDAL, or if GDAL via `ogr2ogr` simply has a few more smarts to handle setting the correct internal spatial domain by inspecting the features as it's creating the FileGDB.
## FileGDB Layer Creation Options

_This may be worth a separate issue, but for context I'll note it here._

Having gotten this far I was hoping to test my theory further by passing the **XORIGIN, YORIGIN, and XYSCALE** layer creation options[2] available in GDAL's FileGDB driver in order to define a grid sufficient to hold all of the features (essentially, the extent of Western Australia).

According to #90 I should just be able to pass them to to `fiona.drivers()`, but this appears to have no effect when it comes to FileGDBs and / or these specific layer creation options.

``` python
import fiona
with fiona.drivers(GEOMETRY_NAME="SHAPEY", FID="OID"):
        with fiona.collection(os.path.join(TMP_DIR, "filegdb-bug/first-geom-wa-extent.json"), "r") as source:
            print source.meta
            with fiona.open(os.path.join(TMP_DIR, "test.gdb"), "w", driver="FileGDB", schema=source.meta["schema"], crs=source.meta["crs"]) as sink:
                for f in source:
                    print f
                    sink.write(f)
```

The resulting FileGDB run through `ogrinfo` reports the defaults for geom name and fid are still in use:

```
FID Column = OBJECTID
Geometry Column = SHAPE
```

Fiona 1.7.0-post2, GDAL 2.1.1, Python 2.7

[1] http://webhelp.esri.com/arcgisdesktop/9.2/index.cfm?id=2059&pid=2057&topicname=Setting_spatial_indexes
[2] http://www.gdal.org/drv_filegdb.html (See the _Important_ note under "XORIGIN, YORIGIN, ZORIGIN, XYSCALE, ZSCALE".)

**Comment by sgillies** [comment]
@keithamoss thanks for the detailed report! I don't see any sign of ogr2ogr changing the default grid options (XORIGIN, YORIGIN, ZORIGIN, XYSCALE, ZSCALE), so the fio commands would be using the same ones. I also don't see any sign of ogr2ogr sorting GeoJSON features by size. OGR will scan your GeoJSON file once before getting to work in order to build a layer definition and therefore would have the extents of the GeoJSON file that fio-load would not, but I don't see where that is passed to the FileGDB layer.

Would you mind running fio-load again with verbosity `-vvv`? This will surface GDAL debug messages that might reveal something.

`GEOMETRY_NAME="SHAPEY"` and `FID="OID"` have no effect because these are layer creation options, not global configuration options. They must be passed as keyword arguments to `fiona.open()`. This reminds me that fio-load doesn't take `--co` options like, e.g., rio-convert and that's something I need to address soon.

**Comment by keithamoss** [comment]
Thanks :)

Here's the output of fio-load with `-vvv` for the three test cases

https://gist.github.com/keithamoss/a3f59096070e648604187665bdb40136

For the sake of comparison I've also included the output of `ogr2ogr` in debug mode for the failing GeoJSON file (small-geom-first.json).

Re passing layer creation options to `fiona.open()` - I've adapted the example above.

```python
with fiona.open(os.path.join(TMP_DIR, "test.gdb"), "w", driver="FileGDB", schema=source.meta["schema"], crs=source.meta["crs"], GEOMETRY_NAME="SHAPEY", FID="OID") as sink:
    for f in source:
        print f
        sink.write(f)
```

only to receive `TypeError: open() got an unexpected keyword argument 'GEOMETRY_NAME'`.

From poking around in Fiona's code I [can see where `fiona.collection` would pass `kwargs` through](https://github.com/Toblerity/Fiona/blob/master/fiona/collection.py#L157), but nothing like that in the newer (?) `fiona.open` approach.

**Comment by sgillies** [comment]
@keithamoss thanks for looking into the keyword argument situation. I'm working on a fix for this in PR #398.

**Comment by keithamoss** [comment]
Thanks a lot @sgillies - I've worked around this for now by outputting GeoJSON and calling `ogr2ogr` to convert to FileGDB directly from Python-land. Will give this change a go and test my theory re pre-calculating XORIGIN, YORIGIN, and XYSCALE Fiona 1.7.2 ships.
